### DISCUSSION TITLE
【v3.7.0-RC】OpenAi 翻译提示词模板、自定义模型等新功能，欢迎抢先体验【已过时】

### DISCUSSION THREAD

**Discussion by YiiGuxing** [YiiGuxing/TranslationPlugin#6235] [Versions]
**_【已过时】v3.7.0 已正式发布_**
----

v3.7 版本的候选版本已经完成开发，此版本包含以下变化：
- 新增 OpenAI 翻译的提示词模板支持
- 新增 OpenAI 翻译的自定义模型支持
- OpenAI 翻译和 TTS 支持自定义 API 路径
- 为 OpenAI TTS 引入了独立的 API 配置

欢迎大家抢先体验：[TranslationPlugin-3.7.0-RC.2.zip](https://github.com/user-attachments/files/18820550/TranslationPlugin-3.7.0-RC.2.zip) （最后更新：2025/2/17）

## 翻译提示词模板配置

此版本新增了对 OpenAI 翻译提示词模板的支持，并通过模板文件的方式提供配置。目前还不支持通过操作UI的方式配置，这个在将来的版本肯定会加上的，请放心。

 翻译提示词模板配置文件有常规翻译提示词模板和文档翻译提示词模板两个文件，模板文件目录：
- Windows: `%LOCALAPPDATA%\Yii.Guxing\TranslationPlugin\openai\`
- 其他: `$XDG_DATA_HOME/Yii.Guxing/TranslationPlugin/openai/` 或 `~/.TranslationPlugin/openai/`

翻译提示词模板文件：
- 常规翻译：`translator.prompt`
- 文档翻译：`document.prompt`

模板文件默认不会生成，如果需修改，在对应位置添加模板文件并编写提示词即可，否则将使用默认提示词。

翻译提示词模板文件格式如下：
```
[ROLE1]
Message 1
[ROLE2]
Message 2
...
```

一个模板文件中至少需要包含一条消息，一条消息又包含消息角色和消息体两个部分。

### 消息角色
通过 `[角色名]` 指定消息的角色，目前支持 `SYSTEM`，`USER` 和 `ASSISTANT` 三个角色。模板中的第一条消息可以省略角色定义，如果第一条消息未指定角色，则该消息默认为 `USER` 角色。
**注意**：角色定义必须独占一行，由 `[` 开始，`]` 结束，并且 `[` 前和 `]` 后不能有其他字符，空白字符也不行。

### 消息体
任何在两个角色定义之间的内容或者角色定义到文件结尾（中间不存在角色定义）的内容都被视为消息体，包括空行。消息体有效内容的前后空行或空白行将被忽略（有效内容内部的空行或空白行会被保留）。
另外消息体所有行的公共缩进也会被忽略，因此如果你的消息体里有和消息角色定义一样的内容时，为了避免其被解析为消息角色定义，你可以为消息体的每一行添加相同的缩进：
```
[USER]
  Message line 1
  [SYSTEM]
  Message line 3
  ...
```
最终它会被渲染为：
- 消息角色：USER
- 消息体：
  ```
  Message line 1
  [SYSTEM]
  Message line 3
  ...
  ```

### 模板引擎
翻译提示词模板使用 [Apache Velocity](https://velocity.apache.org/engine/devel/user-guide.html) 模板引擎进行渲染。因此你可以在模板中使用 Apache Velocity 模板语言进行编写。

#### 模板变量

| **变量** | **说明** |
|:-------:|-----------|
| ${DS} | 符号 `$`，此变量用于转义 `$` 字符 |
| ${UTILS} | 工具对象，集成了一些有用的工具函数，见 [Utils](https://github.com/YiiGuxing/TranslationPlugin/blob/v3.7.0-alpha.1/src/main/kotlin/cn/yiiguxing/plugin/translate/trans/openai/prompt/template/Utils.kt) |
| ${LANGUAGE} | 语言枚举类，提供语言枚举和一些工具函数，可通过 `${LANGUAGE.语言名}` 获取指定语言枚举对象（如：`${LANGUAGE.CHINESE}`），见 [LanguageEnum](https://github.com/YiiGuxing/TranslationPlugin/blob/v3.7.0-alpha.1/src/main/kotlin/cn/yiiguxing/plugin/translate/trans/openai/prompt/template/LanguageEnum.kt) 和 [Lang](https://github.com/YiiGuxing/TranslationPlugin/blob/v3.7.0-alpha.1/src/main/kotlin/cn/yiiguxing/plugin/translate/trans/Languages.kt#L28) |
| ${TEXT} | 要翻译的文本，如果是文档翻译，将是 HTML 格式的文本 |
| ${SOURCE_LANGUAGE} | 翻译文本的原语言枚举对象，见 [Lang](https://github.com/YiiGuxing/TranslationPlugin/blob/v3.7.0-alpha.1/src/main/kotlin/cn/yiiguxing/plugin/translate/trans/Languages.kt#L28)（注：可能为 `AUTO`，需要特别处理） |
| ${TARGET_LANGUAGE} | 翻译目标语言枚举对象，见 [Lang](https://github.com/YiiGuxing/TranslationPlugin/blob/v3.7.0-alpha.1/src/main/kotlin/cn/yiiguxing/plugin/translate/trans/Languages.kt#L28) |

### 默认提示词模板
translator.prompt
```
[SYSTEM]
You are a translator.
The user will provide you with text in triple quotes.
Translate the text#if($LANGUAGE.isExplicit($SOURCE_LANGUAGE)) from ${SOURCE_LANGUAGE.languageName} to#else into#end ${TARGET_LANGUAGE.languageName}.
Do not return the translated text in triple quotes.

[USER]
"""
$TEXT
"""
```
document.prompt
```
[SYSTEM]
You are a html document translator.
The user will provide you with an html document.
Translate the html document#if($LANGUAGE.isExplicit($SOURCE_LANGUAGE)) from ${SOURCE_LANGUAGE.languageName} to#else into#end ${TARGET_LANGUAGE.languageName}.
Do not translate the content inside "pre" and "code" tags.

[USER]
$TEXT
```

*如果大家在体验的过程中遇到任何问题，或者有什么建议，欢迎评论提出。*

**Comment by javaController123** [comment]
![Snipaste_2024-11-27_11-09-59](https://github.com/user-attachments/assets/58f7fafb-f771-4ca8-93c0-d85cb978294d)
是我导入的问题吗？

**Reply by javaController123** [reply]
3.7.0是不是不支持2023.3的idea版本？

**Reply by YiiGuxing** [reply]
你是不是解压后装的？插件安装是不能解压安装包的。

**Reply by javaController123** [reply]
> 你是不是解压后装的？插件安装是不能解压安装包的。

好的，我再试试，有可能是这原因

**Comment by lavender010101** [comment]
您好，后续可以支持自定义完整url吗，有些平台的适配路径格式不太一样，后面可能不是 `/v1/chat/completions`，像智谱AI的完整url `https://open.bigmodel.cn/api/paas/v4/chat/completions`

**Reply by YiiGuxing** [reply]
v3.7.0 正式版会支持

**Reply by YiiGuxing** [reply]
RC版本已经支持并更新，欢迎抢先体验，顺便帮忙测一测，感谢！

**Reply by lavender010101** [reply]
感谢您的回复，经过测试，会显示 “无效的API端点” 错误。很多兼容接口是与 `https://api.openai.com/v1` 这一段兼容，智谱AI中是用 `https://open.bigmodel.cn/api/paas/v4`，OPENROUTER 是 `https://openrouter.ai/api/v1`。希望是把url拆分为 `https://api.openai.com/v1` 和 `/chat/completions` 这样，或者可以有选择让输入框内的url作为 ”Full URL“ 还是 ”Base URL“，然后可以直接输入完整url，就是不知道会不会和您原先的设计有冲突。
![image](https://github.com/user-attachments/assets/d48f2445-15bd-4a0d-8013-aa5483591856)

**Reply by YiiGuxing** [reply]
第一个框只输入 URL 的主机域名部分，不包含路径，第二个框只输入 URL 的路径部分，你把 /api/paas/v4 移到第二个框就可以了，之所以这样做是很多代理都只修改域名即可，而且前半部分还可以和 TTS 的共用。

**Reply by lavender010101** [reply]
好的，了解了。

**Comment by suikastar** [comment]
参考 https://ai.google.dev/gemini-api/docs/openai?hl=zh-cn 填了
```
gemini-1.5-flash
https://generativelanguage.googleapis.com
/v1beta/openai/
key*****************************************
```
不能用

**Reply by YiiGuxing** [reply]
出了什么问题？

**Reply by suikastar** [reply]
这个翻译插件有错误日志吗？在哪个路径下呢？有的话我发日志给你。

**Reply by YiiGuxing** [reply]
你出现问题的现象是什么？没有任何信息我无法初步判断你到底遇到了什么样的问题。

**Reply by suikastar** [reply]
![image](https://github.com/user-attachments/assets/0e38f6da-bcb7-4318-b6c5-abf749efc579)
![image](https://github.com/user-attachments/assets/c780eae9-41ab-42f0-adb5-6fbdb2855191)

**Reply by YiiGuxing** [reply]
你只填了基础的 API URL ，API 的路径你还没有填完整，路径部分后还得加上：v1/chat/completions

看来我把第一个框限制到只能填到域名部分还是有些问题的，有一定的误导性，后面我修改一下吧。

**Reply by suikastar** [reply]
~~尝试把 `/v1beta/openai/` 改成了 `/v1beta/openai/v1/chat/completions` 还是同样的报错，不知插件是否有日志，我把日志发你排查一下，或者作者大大您亲自去谷歌免费领取一个Gemini api密钥，自己试试~~

~~使用`/v1beta/openai/v1/chat/completions` ，重启IDE后，可以用了，感谢回复，已经解决问题了~~

好像是命中缓存了，刚试了一下，还是不行

**Reply by YiiGuxing** [reply]
错误是不会被缓存的，还是 not  found 吗？

**Reply by suikastar** [reply]
嗯 not found。翻译了新的地方的注释，可能被以前缓存过。

**Reply by YiiGuxing** [reply]
那你清除一下磁盘缓存（在插件设置页清除），再重启 IDE 试试

**Reply by suikastar** [reply]
试了，把缓存清空了，IDE重启了，还是不行。换硅基流动 api立马行，Gemini不知道为什么不行。

**Reply by YiiGuxing** [reply]
那不应该啊，not found 说明 API 地址不对，你再检查检查，看看前后没有没多个空格什么的

**Reply by YiiGuxing** [reply]
@suikastar 你填的地址不对，两个框的内容（包括默认值）连接起来要和目标URL一致才行。如：
![image](https://github.com/user-attachments/assets/0c6da5c6-6111-429c-b1f6-d958a82ab525)
图中两个框虽然是空的，但它们的默认值拼起来后就是官方的地址：`https://api.openai.com/v1/chat/completions`

而 gemini API 的地址是：`https://generativelanguage.googleapis.com/v1beta/openai/chat/completions` 见：https://ai.google.dev/gemini-api/docs/openai?hl=zh-cn#rest （一开始咱们都没找对）

所以第一个框你要填 `https://generativelanguage.googleapis.com`，而第二个框是路径部分：`/v1beta/openai/chat/completions`

**Reply by suikastar** [reply]
@YiiGuxing 太对了哥，细还是你细，这下可以正常使用了
